### ISSUE TITLE
Points progress bar

### ISSUE THREAD

**Issue by mtthwhggns** [apluslms/a-plus-design-system#3]

**Comment by mtthwhggns** [comment]
I just had a discussion with @saarasat , as she has been working on the questionnaires, which use lots of these.

We agreed to create a proof-of-concept for the replacement of both the progress bars and the badges (#4) with a single component which works at lots of sizes. 

It would be an icon (often a pie-chart), followed by either a number in the current badge format or one or two words.

For example:

Icon | Example Text | Description
-----|-------|---------
Ellipsis inside a circle | 10 | There are 10 points available and this hasn't been attempted yet.
Empty pie chart | 0/10 | The student made an attempt and got zero points from this.
Part-filled pie-chart | 6/10 | The student got more than zero and less than full points - the pie represents what proportion they got full points in.
Full pie chart with an icon inside (maybe a star or tick) | 10/10 | Full points - the icon makes it easier to tell apart from high numbers at small sizes.
Clock | Waiting | This hasn't been graded yet (maybe it's Rubyric, or slow automated grading).
Exclamation mark in a circle | Error | Something went wrong with the grading.

At bigger sizes (like the sidebar), the big icon and text can be arranged vertically. 

Here are some poor-quality doodles we did:

![Points icon doodles](https://user-images.githubusercontent.com/6674046/88167075-efa7e900-cc20-11ea-9f90-5f3c54459a22.png)

One of us will create a proof-of-concept to try out properly, but feel free to comment with any insights/missed use cases here in the meantime.

**Comment by etanttila** [comment]
How will the progress within a module be displayed? Those probably work for single exercises (small) and the entire course (large, on the side), but how about within a module. Personally I think it would be nice if there was an indicator larger than the points badge there as well because it allows one to quickly see one's progress on the module (and whether its done, in progress or not started yet) like the progress bar there currently.

**Comment by mtthwhggns** [comment]
> How will the progress within a module be displayed? Those probably work for single exercises (small) and the entire course (large, on the side), but how about within a module. Personally I think it would be nice if there was an indicator larger than the points badge there as well because it allows one to quickly see one's progress on the module (and whether its done, in progress or not started yet) like the progress bar there currently.

To show progress within a module/chapter, we currently use a badge in the panel header and a very long progress bar inside the panel body (if I've missed any, please let me know). I think the options are:

1. We don't need both of these so close to one another, one (probably the one in the panel header) is enough
2. We rearrange the content inside the panel so that we can fit one of these indicators beside the deadlines, instead of below as it is now. This would change the shape a little. 

My thinking was that this component wouldn't have any fixed size, but would adapt to the side of the text it is displayed around. This would be better for accessibility than the current situation, and would make the component much more versatile. 

But these are just my ideas, feel free to share any more.

**Comment by jaguarfi** [comment]
Regarding the **waiting **message. I think that the current **Grading** message is more meaningful. 

One question. are you planning to show both, the number of submissions and the points ? or are you planning to merge those two into one?

**Comment by mtthwhggns** [comment]
> Regarding the **waiting **message. I think that the current Grading message is more meaningful.

Is there a useful distinction to be made between "waiting" (possibly in a queue, possibly waiting for manual grading in Rubyric or whatever) and "grading" (something is actually happening now, probably automated)? An active term like "grading" feels misleading if it's going to take two weeks for a TA to look at it and grade it manually. 

> are you planning to show both, the number of submissions and the points ? or are you planning to merge those two into one?

This idea has them separately and only touches the points indicator, but, again, feel free to share thoughts on whether they could be meaningfully combined.

**Comment by jaguarfi** [comment]
> "waiting" (possibly in a queue, possibly waiting for manual grading in Rubyric or whatever) 

I mean that is exactly the point, you can wait for **whatever**. Students do not know what is happening behind the submission,  and they're most likely not interested in knowing. They only want to submit something and get a grade as soon as possible. Therefore, if they see a waiting message, they could become confused because, in fact, the students are the ones waiting for something to happen. Now, with the waiting message on the screen they can start thinking that the system is waiting for a response from somewhere on the internet (db, server) or maybe the system is waiting for some input from the student.

For me, the word grading at least tell the user what is happening. Indeed, the word grading does not say how long the grading is going to take, but at least it tells the students what is happening in the system and implicitly tell the students that they should wait for the grade.

I mean this is more a semantic discussion :nerd_face:  and you are the native speaker, so you might be right in your logic. But for me as a non-native speaker the word waiting can be confusing :zany_face:

>This idea has them separately and only touches the points indicator, but, again, feel free to share thoughts on whether they could be meaningfully combined.

I think having both is good, and I like the new concept :+1:

**Comment by mtthwhggns** [comment]
Taking the comments so far into account, it's time for a better mockup.

Don't pay any attention to the size. I'm creating SVGs which should look good at any size these can be used, I just wanted them to be big enough to see clearly in this example. 

Also, don't pay attention to the colours. They'll be read from the SASS functional colours. 

Icon | Possible text | Notes 
----------|---------|---------
![not_started](https://user-images.githubusercontent.com/6674046/89867759-56824780-dbba-11ea-9107-d129a826bec9.png) | 12 points | This exercise isn't started yet, but it is worth up to 12 points. 
![empty](https://user-images.githubusercontent.com/6674046/89867891-8af60380-dbba-11ea-8719-f96b4962332f.png) | 0/12 points | You attempted this exercise, but got zero points.
![partial](https://user-images.githubusercontent.com/6674046/89867932-9b0de300-dbba-11ea-8afd-58134f7071b8.png) | 9/12 points | You attempted this exercise and got 9/12 possible points.
![full](https://user-images.githubusercontent.com/6674046/89867962-aa8d2c00-dbba-11ea-85d0-a6957461aa40.png) | 12/12 points | You attempted this exercise and got full points.
![rejected](https://user-images.githubusercontent.com/6674046/89868020-c5f83700-dbba-11ea-9e73-3ceb0ec72611.png) | Rejected | Submission was rejected
![error](https://user-images.githubusercontent.com/6674046/89868108-f2ac4e80-dbba-11ea-9a22-86b543dc71b2.png) | Error | An error occurred when attempting to grade this
![delayed](https://user-images.githubusercontent.com/6674046/89868233-22f3ed00-dbbb-11ea-8726-05ffbd100d7c.png) | Grading | Either this is being graded right now (automated grading) or it's waiting to be graded by a human. 

I'll be doing some more work on these today, so any thoughts or feedback would be appreciated as soon as possible!

**Comment by etanttila** [comment]
I think those mostly look really good. Personally I might hope the star in the full points would be a little bit smaller so even when the icon is small, it would be clear that it's a filled circle with an icon (and not just something with an outline). But I think it shouldn't be much smaller, because it's good that it's a different size than the rejection X. It's currently a bit off center as well (on the vertical axis), I'd prefer it be in the center.

**Comment by etanttila** [comment]
Are the "Notes" draft versions of tooltips that appear when hovering the mouse on the icon/points?

**Comment by mtthwhggns** [comment]
>  It's currently a bit off center as well (on the vertical axis), I'd prefer it be in the center.

~~It's centered to the geometric centre of the star - look at a square added around it~~ It turns out this isn't the "geometric centre". Oops. 

![Näyttökuva 2020-8-11 kello 10 42 24](https://user-images.githubusercontent.com/6674046/89871062-79fbc100-dbbf-11ea-9b4f-ee60b1bb028d.png)

There aren't any specific plans for tooltips yet, the Notes were just meant to explain what it's meant to mean.

**Comment by raphendyr** [comment]
> You attempted this exercise and got 9/12 possible points.

Would there be versions for every quarter, 10%, or some other number of possible partial points?

> This exercise isn't started yet, but it is worth up to 12 points.

Three dots might look like a button to get options. but it's also possible that the symbol works better with the color.

**Comment by mtthwhggns** [comment]
> Would there be versions for every quarter, 10%, or some other number of possible partial points?

Yes. The SVG shows any proportion as a pie chart, as originally described.

**Comment by etanttila** [comment]
> > It's currently a bit off center as well (on the vertical axis), I'd prefer it be in the center.
> 
> It's centered to the geometric centre of the star - look at a square added around it
> 

It seems like by the geometric center you mean here the midpoint of the square, which actually isn't the geometric center of the star itself. It's probably the easiest to place it that way, although a person usually doesn't imagine a square around the icon but rather compares how close the points are to the edges (and here how much black there is in each part). You'll notice that it's clear that the star itself (rather than the box containing it) is off center if you look at the icon at a 112 or 90 degree angle.

EDIT: illustration
![image](https://user-images.githubusercontent.com/50318434/89874964-49b72100-dbc5-11ea-8045-b552cbc7e96a.png)

**Comment by mtthwhggns** [comment]
Thanks @etanttila , I did not consider this properly at all.

![Näyttökuva 2020-8-11 kello 11 14 40](https://user-images.githubusercontent.com/6674046/89874027-f1cbea80-dbc3-11ea-8f40-f41e9c5a4523.png)

Is this better?

**Comment by etanttila** [comment]
It's great on center! The size could be good, but I think the star could be slightly larger (so it's a different size than the rejected X), so perhaps somewhere between the two.

**Comment by markkuriekkinen** [comment]
I like the new SVG icons. It would be nice to see a mock-up of how they really look inside A+ (but let's not go there yet if it is too laborious to set up, depending on how you progress with this work). The module total points bar is indeed a bit different case from the other points badges and requires some designing or experimentation. I wonder if these pie chart icons will be large enough so that people really can differentiate the points percentage visually. The old points progress bars have always been clear since they take a lot space on the screen. Overall, creating this new progress/badge combination is a great idea!

About the word "waiting" vs. "grading": some years ago, the submission status "waiting" was renamed in the user interface to "in grading" or was it "waiting for grading" so that students do not think that the system is waiting for them to input something. Using the word "grading" implies that the system is waiting for the grading to finish and the student need not do anything more.

**Comment by mtthwhggns** [comment]
I spent today getting it implemented in A+. Here's the results so far for the small (just slightly taller than text) height:

![localhost_8000_accounts_login_ (3)](https://user-images.githubusercontent.com/6674046/89889738-d5877800-dbda-11ea-8d62-7563d0aedcf1.png)

The colours come from the brand variables in SASS and the percentage is calculated automatically.

**Comment by raphendyr** [comment]
Error is usually problem with the sustem, which the user can't do anything about. Ideally, it should solve itself without any action from the student, thus it has been gray. I'm not sure what would be most correct behavior there.

**Comment by mtthwhggns** [comment]
> About the word "waiting" vs. "grading": some years ago, the submission status "waiting" was renamed in the user interface to "in grading" or was it "waiting for grading" so that students do not think that the system is waiting for them to input something. Using the word "grading" implies that the system is waiting for the grading to finish and the student need not do anything more.

To me, it seems like very bad UX if we don't indicate to users whether "the thing is being graded now and there will be results in 20-30 seconds, you probably want to wait and see the results before you move on in case you need to fix something" or whether "the thing has been sent to Rubyric and will be manually graded in 2-3 weeks. Please don't sit there expecting points to appear". 

I think we should probably have a way to set those expectations at some point, but it's not urgent now. 

> Error is usually problem with the sustem, which the user can't do anything about. Ideally, it should solve itself without any action from the student, thus it has been gray. I'm not sure what would be most correct behavior there.

Can we rely on it being used this way or is it a status that grader authors can choose to return? If it's reliable, I agree that grey is good and I'm happy to swap it.

**Comment by raphendyr** [comment]
> Can we rely on it being used this way or is it a status that grader authors can choose to return? If it's reliable, I agree that grey is good and I'm happy to swap it.

The problem would be between A+ and grader (e.g. network down or grader down). There is a long goal to fix this so that it would be fixed automatically, so I would go with gray. The relevant issue https://github.com/apluslms/a-plus/issues/470

**Comment by mtthwhggns** [comment]
It's possible to have modules with no points available, here's an example from `course_template`:

![localhost_8000_def_current_ (1)](https://user-images.githubusercontent.com/6674046/89985626-0cfe2f00-dc84-11ea-8abf-a380876f1e34.png)

It seems misleading to show the "full points" style here (although we do that currently with the coloured badges). Compared to the others;

![localhost_8000_def_current_ (4)](https://user-images.githubusercontent.com/6674046/89985964-7ed67880-dc84-11ea-9a5d-8d1a44f47fff.png)

![localhost_8000_def_current_ (3)](https://user-images.githubusercontent.com/6674046/89985820-49318f80-dc84-11ea-9d05-1e8ae4356a37.png)

this feels out of place. 

How do you feel about showing text instead (something like "no points available"? Or should there be an icon specifically for this scenario?

**Comment by raphendyr** [comment]
I think there could be icons for those cases. There has been plan to replace point info for exercises with a textual representation, when the exercise max points are 2 or less (see https://github.com/apluslms/a-plus/issues/488).

**Comment by etanttila** [comment]
I think the "full points" style is very misleading especially if there are some exercises worth zero points and the user hasn't attempted them. I think the text could work or optionally a specific icon if one can come up with a good one.

**Comment by markkuriekkinen** [comment]
> It's possible to have modules with no points available

Perhaps text "no points available" is much easier to understand than having a lot of unique icons for all the rare, obscure cases.

**Comment by raphendyr** [comment]
For modules, "no points available" could be fine. However, I don't know if it adds any information. Maybe there just shouldn't be anything at the end of the line for that module. Would that be too confusing?

**Comment by etanttila** [comment]
> > It's possible to have modules with no points available
> 
> Perhaps text "no points available" is much easier to understand than having a lot of unique icons for all the rare, obscure cases.

I agree with this. However, I also think that we probably need to think should there be an icon for 0/0 exercises (and how those are displayed), and if there is a separate one for those, the same could be used for modules, because then 0/0 in a module isn't an obscure case but rather just another example of the 0/0 instances. Because 0/0 exercises aren't obscure cases but common (at least in Programming 1).

**Comment by mtthwhggns** [comment]
There seem to be at least two distinct `/0` cases: 

- O1 has them as "optional" or "practice" exercises - you can pass or fail them, but they don't affect your grade. 
- Programming Studio 2 used them for "choose a project" or "submit an assignment that's manually graded" - these were worth a huge proportion of the grade, but the points were never shown in the A+ interface. I guess they lived somewhere else. The actual points were a Rubyric comment.

**Comment by etanttila** [comment]
> For modules, "no points available" could be fine. However, I don't know if it adds any information. Maybe there just shouldn't be anything at the end of the line for that module. Would that be too confusing?

I think it would be good to have at least the 0/0 at the end of the line, (or if that isn't wanted there, a text), text or icon may not be mandatory, but it would e good to have at least something that indicates that there aren't any points available

**Comment by etanttila** [comment]
> There seem to be at least two distinct `/0` cases:
> 
>     * O1 has them as "optional" or "practice" exercises - you can pass or fail them, but they don't affect your grade.
> 

And I suppose in this instance "no points" and "full points" are good.

>     * Programming Studio 2 used them for "choose a project" or "submit an assignment that's manually graded" - these were worth a huge proportion of the grade, but the points were never shown in the A+ interface. I guess they lived somewhere else. The actual points were a Rubyric comment.

And here if one has a submission for choosing a project, it would be "full points" as well. The other manually graded assignment is more difficult to categorize.

So I suppose if a module didn't have any exercises, it would be different from either of these instances. So Markku may be right that modelus without points could be an obscurish case, if they're not really "rinnastettavissa" to the exercises

**Comment by ttsirkia** [comment]
I think the first icon "This exercise isn't started yet" is not very recognizable and looks more like a button to open a menu. I cannot explain the meaning if I don't know it by heart.

Please also consider that icons still have colors. Although it has a symbol, making the difference with a quick glance is faster with colors.

Microsoft tried to get rid of all colors a few years back and the feedback was massive: https://devblogs.microsoft.com/visualstudio/introducing-the-new-developer-experience/ and https://devblogs.microsoft.com/visualstudio/visual-studio-11-user-interface-updates-coming-in-rc/

**Comment by mtthwhggns** [comment]
> I think the first icon "This exercise isn't started yet" is not very recognizable and looks more like a button to open a menu. I cannot explain the meaning if I don't know it by heart.
> 
> 
> 
> Please also consider that icons still have colors. Although it has a symbol, making the difference with a quick glance is faster with colors.
> 
> 
> 
> Microsoft tried to get rid of all colors a few years back and the feedback was massive: https://devblogs.microsoft.com/visualstudio/introducing-the-new-developer-experience/ and https://devblogs.microsoft.com/visualstudio/visual-studio-11-user-interface-updates-coming-in-rc/

I'd welcome any suggestions for a replacement for the first one. I agree that it looks like a "kebab menu" - I was going for an ellipsis. 

None of this is finished yet, I didn't reach a conclusion on colour yet. I was also experimenting with different saturations of the brand colour.

**Comment by ttsirkia** [comment]
![kuva](https://user-images.githubusercontent.com/2797729/90776309-9ce64d80-e302-11ea-8cc9-4b2056682132.png)

The picture is not perfect but presents some ideas.

If the title bar of the exercise rounds had a white section, there would be a possibility to use icons with colors.

The icon with three dots is replaced with an empty circle which is gray. The amount of points is in parenthesis to make the difference with gained zero points after a submission.

As stars are often used with favorites etc. I replaced the symbol with a tick. Also the clock icon is replaced with an hour glass which indicates better an action to wait.

There are textual explanations for the error cases. The usage of those pies would require at least three icons: 25 %, 50 % and 75 %.

**Comment by ttsirkia** [comment]
There is still the discussion about the rounds or exercises in which the number of points is below what is required. How that should be indicated and is the symbol always red in those cases?

**Comment by raphendyr** [comment]
> The icon with three dots is replaced with an empty circle which is gray. The amount of points is in parenthesis to make the difference with gained zero points after a submission.

There should be more distinguishable symbol for not started round (not required for submissions). In addition, when there is no submission, then there is no 0 points.

However, this pointed out that we have to handle unofficial points. For that, we can use gray for the color and also something on the symbol side. As the circle is used for the percentage part, we could use parenthesis, but that might not be clear enough.

> As stars are often used with favorites etc. I replaced the symbol with a tick. Also the clock icon is replaced with an hour glass which indicates better an action to wait.

Personally, I think the star is better. There is no "correct" for an exercise, only for sub parts of the exercise as 11/12 is correct as well as 12/12. The star can be used to highlight that this is the "best" score. Though, sparkles (:sparkles: ) might be better symbol for that, though one star should be more clear (less small things).

> There are textual explanations for the error cases. The usage of those pies would require at least three icons: 25 %, 50 % and 75 %.

There is an implementation using SVG to calculate the correct fillness for the circle, so we be exact.

Finally, when there is "Error", then no submission was consumed and no points are given for that submission, hence we shouldn't render `0 / n` there, just the text.

**Comment by etanttila** [comment]
I agree that the ellipsis is not ideal for "not started yet", as it is often used with menus, but I don't really have better ideas. I kind of like the tick better than the star, although I understand why some would prefer the star.

**Comment by ttsirkia** [comment]
If there are no submissions, it could also show "max 5 p." instead of "0 / 5" or "(0 / 5)" as the user has not received any points yet.

**Comment by markkuriekkinen** [comment]
The hourglass for the "waiting for grading" status icon is a nice idea. It does not get mixed with the pie chart points icons, while the clock icon earlier suggested for the waiting status looks a little similar to the points pie charts. The hourglass is used for "waiting" in many other places too.

I agree with Teemu that colourful icons are good for those with normal colour vision.

Using the empty circle for "no submissions" looks very similar to "submissions with zero points", and Matthew probably aimed at making distinct icons that do not depend on colour to differentiate them. What would be a common icon for "a task on the TODO list waiting for you to get started"?

Using the tick icon instead of the star for "full points" is also a good idea. The tick icon is more widely used in this context (a completed task). Jaakko is right in that the students are not always required to gain full points and the task could be finished with less than max points, but we can never know if the student wants to keep trying until he/she has full points. I wonder if it looks a little childish if we are giving students "stars" every time they get the full score. (Think of the Finnish "papukaijamerkki" at schools; do we want to give funny badges at the university to students with full scores?) The check mark/tick is more neutral: just a completed task.

When SVG images are used, is there any performance or scalability issue when the page includes dozens of the SVG icons?

**Comment by ttsirkia** [comment]
> When SVG images are used, is there any performance or scalability issue when the page includes dozens of the SVG icons?

That shouldn’t be any problem for modern browsers. Web sites are nowadays full of svg icons.

**Comment by ttsirkia** [comment]
> What would be a common icon for "a task on the TODO list waiting for you to get started"?

I have been thinking this without any better ideas. In my opinion the best option at the moment would be to show the empty (possibly gray) circle and only the maximum amount of points that can be achieved. When the zero is not shown, it cannot be interpreted as gained points. After first submission the received points (zero or more) are shown.

Of course if there was a meaningful icon for this, it would be nice.

**Comment by raphendyr** [comment]
> Using the empty circle for "no submissions" looks very similar to "submissions with zero points", and Matthew probably aimed at making distinct icons that do not depend on colour to differentiate them. What would be a common icon for "a task on the TODO list waiting for you to get started"?

One possibility would be using a dashes line for the circle or something other effect, which makes it not empty circle. I think moodle used boxes with a dashed line for something too.

> Using the tick icon instead of the star for "full points" is also a good idea. The tick icon is more widely used in this context (a completed task).

This would make even more sense if we made it possible to mark tasks complete manually. I.e. once the users feels they are completed with a task, they would be able to press a button "I'm done with this", which would help them to journal the work. Additianlly, that could lock the exercise (or exercises for a module), when the task is complete and would allow showing the model for example or such.

Assuming this manual marking feature exists, then it would make sense to mark exercises and modules as _complete_, as then it would really mean that and not just "hey, you got all points!". Sadly, addind this feature wouldn't be so easy, but we can assume it to exists from the design perspective of this component.

**Comment by etanttila** [comment]
> > Using the empty circle for "no submissions" looks very similar to "submissions with zero points", and Matthew probably aimed at making distinct icons that do not depend on colour to differentiate them. What would be a common icon for "a task on the TODO list waiting for you to get started"?
> 
> One possibility would be using a dashes line for the circle or something other effect, which makes it not empty circle. I think moodle used boxes with a dashed line for something too.

A dashed circle could be a possible option. 

> This would make even more sense if we made it possible to mark tasks complete manually. I.e. once the users feels they are completed with a task, they would be able to press a button "I'm done with this", which would help them to journal the work. Additianlly, that could lock the exercise (or exercises for a module), when the task is complete and would allow showing the model for example or such.
> 
> Assuming this manual marking feature exists, then it would make sense to mark exercises and modules as _complete_, as then it would really mean that and not just "hey, you got all points!". Sadly, addind this feature wouldn't be so easy, but we can assume it to exists from the design perspective of this component.

I think this manually marking feature would be unnecessary and not much used. Most users wouldn't realize it even exists, and those that do probably wouldn't use it much. Personally I don't think it would be worth the effort and might even confuse users.

**Comment by ttsirkia** [comment]
This could work.
![kuva](https://user-images.githubusercontent.com/2797729/91022603-1e3d2900-e5fe-11ea-9588-14b2dca01bf6.png)

**Comment by raphendyr** [comment]
> I think this manually marking feature would be unnecessary and not much used. Most users wouldn't realize it even exists, and those that do probably wouldn't use it much. Personally I don't think it would be worth the effort and might even confuse users.

I think this was requested at some point, however I couldn't find an issue about it, so it hasn't gotten so far yet, I think.

**Comment by raphendyr** [comment]
I think we need to think all of these in three different locations:

- as the status for a module (doesn't include states like in grading and error)
- as the status for an exercise (should show the status of the best exercise)
- as the status for a submission, i.e., in the list of all submissions for an exercise.

For the module and exercise, we need to be able to show if the items "points to pass" limit has been reached.

For the submission, we need to handle unofficial submission, i.e., those that are not actually used as they are too late or are over the submission limit.
